tools/syz-declextract/run.go

```diff
@@ -122,6 +151,237 @@ type context struct {
+	probeFuncToFiles    map[string]map[string]bool
+	fdNames             map[string]int
+	ioctlNames          map[string]int
+}
+
+// OutputTop represents json output of the clang tool.
+type OutputTop struct {
+	Fops *OutputFops
+}
+
+// OutputFops describes one file_operations variable.
+type OutputFops struct {
+	// Names of callback functions.
+	Open  string
+	Read  string
+	Write string
+	Mmap  string
+	Ioctl string
+	Cmds  []OutputIoctlCmd // set of ioctl commands
+}
+
+type OutputIoctlCmd struct {
+	Name string // literal name of the command (e.g. KCOV_REMOTE_ENABLE
+	Type string // inferred syzlang type (e.g. ptr[in, int32])
+}
+
+func (fops *OutputFops) String() string {
+	w := new(bytes.Buffer)
+	fmt.Fprintf(w, "fops:")
+	writefop := func(name, fn string) {
+		if fn != "" {
+			fmt.Fprintf(w, " %v:%v", name, fn)
+		}
+	}
+	writefop("open", fops.Open)
+	writefop("read", fops.Read)
+	writefop("write", fops.Write)
+	writefop("mmap", fops.Mmap)
+	writefop("ioctl", fops.Ioctl)
+	return w.String()
+}
+
+func probe(cfg *mgrconfig.Config, cfgFile string, cache bool) (*ifaceprobe.Info, error) {
+	if cache {
+		info, err := readProbeResult(cfg)
+		if err == nil {
+			return info, nil
+		}
+	}
+	_, err := osutil.RunCmd(30*time.Minute, "", filepath.Join(cfg.Syzkaller, "bin", "syz-manager"),
+		"-config", cfgFile, "-mode", "iface-probe")
+	if err != nil {
+		return nil, err
+	}
+	return readProbeResult(cfg)
+}
+
+func readProbeResult(cfg *mgrconfig.Config) (*ifaceprobe.Info, error) {
+	data, err := os.ReadFile(filepath.Join(cfg.Workdir, "interfaces.json"))
+	if err != nil {
+		return nil, err
+	}
+	info := new(ifaceprobe.Info)
+	if err := json.Unmarshal(data, info); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal interfaces.json: %w", err)
+	}
+	return info, nil
+}
+
+func (ctx *context) parseProbeInfo(info *ifaceprobe.Info) {
+	pcToFunc := make(map[uint64]string)
+	for _, pc := range info.PCs {
+		pcToFunc[pc.PC] = pc.Func
+	}
+	ctx.probeFuncToFiles = make(map[string]map[string]bool)
+	for _, file := range info.Files {
+		for _, pc := range file.Cover {
+			fn := pcToFunc[pc]
+			files := ctx.probeFuncToFiles[fn]
+			if files == nil {
+				files = make(map[string]bool)
+				ctx.probeFuncToFiles[fn] = files
+			}
+			files[file.Name] = true
+		}
+	}
+}
+
+func (ctx *context) createFops(fops *OutputFops) {
+	// TODO: also emit interface entry for the fops.
+	if fops.Read == "" && fops.Write == "" && fops.Mmap == "" && fops.Ioctl == "" {
+		return
+	}
+	name, files := ctx.mapFopsToFiles(fops)
+	if len(files) == 0 {
+		fmt.Printf("%v: %v is not mapped to any file\n", name, fops)
+		return
+	}
+	fmt.Printf("%v: %v mapped to %v\n", name, fops, files[:min(10, len(files))])
+
+	// Some fops are mapped to too many files, usually these have generic callbacks
+	// like simple_attr_read+simple_attr_write. Compiler restricts number of strings to 2000.
+	// TODO: emit multiple open calls to cover all files.
+	files = files[:min(len(files), 1000)]
```

Comment:
Maybe take less than 1000?

With so many files, the chances that syzkaller actually tries them all out are very small anyway.